src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalHelpers.cs

```diff
@@ -244,11 +247,17 @@ public static partial class MarshalHelpers
             if (marshalAs != null)
                 nativeType = marshalAs.Type;
 
+            if (isField && nativeType == NativeTypeKind.CustomMarshaler)
+                return MarshallerKind.Invalid;
+
             //
             // Determine MarshalerKind
             //
             if (type.IsPrimitive)
             {
+                if (nativeType == NativeTypeKind.CustomMarshaler)
```

Comment:
How does this case work in CoreCLR? I do not think it works in this implementation.

The `ICustomMarshaler` takes object. The primitive type would have to be boxed to get converted to object, but I do not see any boxing in the generated IL.

Reply:
Primitive types are going to go into the throwing path with rest of IsValueType types.

Instead of putting all these `if (nativeType == NativeTypeKind.CustomMarshaler)` checks everywhere, can we do this like CoreCLR before we start looking in detail at `type`:

https://github.com/dotnet/runtimelab/blob/0aeac63238f2b7cd6b788b66b5fd9655cc183a60/src/coreclr/vm/mlinfo.cpp#L1264-L1327

I.e. do a single check for nativeType == NativeTypeKind.CustomMarshaler 3 lines above this one.

We're still missing handling of CustomMarshaller on e.g. string and delegate below, which I assume is possible with CoreCLR. A single check should accomplish what we need.